test/provider_fallback_test.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stddef.h>
+#include <openssl/provider.h>
+#include <openssl/evp.h>
+#include "testutil.h"
+
+static int test_provider(OPENSSL_CTX *ctx)
+{
+    EVP_KEYMGMT *rsameth = NULL;
+    const OSSL_PROVIDER *prov = NULL;
+    int ok;
+
+    ok = TEST_true(OSSL_PROVIDER_available(ctx, "default"))
+        && TEST_ptr(rsameth = EVP_KEYMGMT_fetch(ctx, "RSA", NULL))
+        && TEST_ptr(prov = EVP_KEYMGMT_provider(rsameth))
+        && TEST_str_eq(OSSL_PROVIDER_name(prov), "default");
+
+    EVP_KEYMGMT_free(rsameth);
+    return ok;
+}
+
+static int test_fallback_provider(void)
+{
+    return test_provider(NULL);
+}
+
+static int test_explicit_provider(void)
+{
+    OPENSSL_CTX *ctx = NULL;
+    OSSL_PROVIDER *prov = NULL;
+    int ok;
+
+    ok = TEST_ptr(ctx = OPENSSL_CTX_new())
+        && TEST_ptr(prov = OSSL_PROVIDER_load(ctx, "default"))
+        && test_provider(ctx)
+        && TEST_true(OSSL_PROVIDER_unload(prov));
+
+    OPENSSL_CTX_free(ctx);
+    return ok;
+}
+
+
+int setup_tests(void)
+{
+    /*
+     * We must ensure that there is no OPENSSL_CONF defined.  Otherwise,
+     * we risk that the configuration file contains statements that load
+     * providers, which defeats the purpose of this test.
+     */
+#ifndef _WIN32
+    unsetenv("OPENSSL_CONF");
```

Comment:
This seems unnecessary:

https://github.com/openssl/openssl/blob/a5a87011baeef71c86938a2bae54f89fbe99e5dc/test/run_tests.pl#L30

Reply:
Some day, someone will make a change. I'd rather be extra safe.

Reply:
> Some day, someone will make a change. I'd rather be extra safe.

And in that event the test will likely fail, and that person will fix it. Should we do this with all our tests? I don't think so.

Reply:
This test is *particularly sensitive* in this regard, as it tests a specific condition (fallbacks).  Not all our tests do that, specifically.

Reply:
There are lots of tests that would start failing under these conditions. I don't think this one is any more sensitive in that regards.

Reply:
Huh?  The other tests might fail if `OPENSSL_CONF` is undefined, this test will most likely not be able to check what it's supposed to if `OPENSSL_CONF` *is **defined***!  How do you not see the difference?

Reply:
For example, what happens if we decide to globally define `OPENSSL_CONF` with `test/default-and-legacy.cnf`?  Will the other tests fail?  What will happen with this one?

Reply:
...

I guess that the other alternative is to ensure I have OPENSSL_CONF defined to the NUL device, in the test recipe.

Reply:
Done

Reply:
> For example, what happens if we decide to globally define OPENSSL_CONF with test/default-and-legacy.cnf? Will the other tests fail?

I thought that was an interesting question, so I tried it out:

````
Test Summary Report
-------------------
20-test_provider.t               (Wstat: 256 Tests: 9 Failed: 1)
  Failed test:  9
  Non-zero exit status: 1
90-test_sslapi.t                 (Wstat: 512 Tests: 3 Failed: 2)
  Failed tests:  1, 3
  Non-zero exit status: 2
Files=198, Tests=1993, 309 wallclock secs ( 4.74 usr  0.50 sys + 295.25 cusr 18.45 csys = 318.94 CPU)
Result: FAIL
````

So, it turns out, that my "lots of tests that would start failing under these conditions" was an exaggeration - although some do. However, I much prefer your new solution.

Reply:
The trouble now is that if someone wants to try that particular test program manually, they must remember to define `OPENSSL_CONF` "properly".

(this is a pet peeve of mine, btw, that when running the tests verbosely, we don't display the values of diverse important environment variables, which makes manual testing a royal pain...  Hmmm, I should probably raise an issue for that)